File: 792.匹配子序列的单词数.go
/*
 * @lc app=leetcode.cn id=792 lang=golang
 *
 * [792] 匹配子序列的单词数
 */
package leetcode

// @lc code=start
func numMatchingSubseq(s string, words []string) int {
	smap := make(map[string]int)
	res := 0
	for _, word := range(words) {
		smap[word]++
	}
	for k, v := range(smap) {
		if isSub(k, s) {
			res += v
		}
	}
	return res
}

func isSub(word string, s string) bool {
	i, j := 0, 0
	for ; i < len(word) && j < len(s); j++ {
		if word[i] == s[j] {
			i++
		}
	}
	return i >= len(word) 
}
// @lc code=end

File: 792.匹配子序列的单词数.cpp
/*
 * @lc app=leetcode.cn id=792 lang=cpp
 *
 * [792] 匹配子序列的单词数
 */
#include <string>
#include <vector>
using namespace std;
// @lc code=start
class Solution {
public:
    int numMatchingSubseq(string s, vector<string>& words) {
        int res = 0;
        for (auto &word : words) {
            if (isSub(word, s)) {
                res++;
            }
        }
        return res;
    }
    bool isSub(string& word, string& s) {
        size_t p = 0;
        for (auto ch : word) {
            p = s.find(ch, p);
            if (p == string::npos) 
                return false;
            else 
                p++;
        }
        return true;
    }
};
// @lc code=end

